### PR TITLE
Add MLFlowLogger

### PR DESCRIPTION
**Pull request by eracah**
# What does this PR do?

Adds basic MLFlowLogger that support the following things:
* setting experiment name
* setting run name
* setting tracking_uri
* logging metrics
* logging params (hyperparameters)

tested manually and added unit tests

# What issue(s) does this change relate to?

fixes CO-1245

# Before submitting
- [x] Have you read the [contributor guidelines](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md)?
- [ ] Is this change a documentation change or typo fix? If so, skip the rest of this checklist.
- [x] Was this change discussed/approved in a GitHub issue first? It is much more likely to be merged if so.
- [x] Did you update any related docs and document your change?
- [x] Did you update any related tests and add any new tests related to your change? (see [testing](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#running-tests))
- [x] Did you run the tests locally to make sure they pass?
- [x] Did you run `pre-commit` on your change? (see the `pre-commit` section of [prerequisites](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#prerequisites))

### CODE REVIEW

File: composer/loggers/__init__.py [hunk 1]
```diff
@@ -18,21 +18,14 @@
...
-    'FileLogger',
-    'InMemoryLogger',
-    'ProgressBarLogger',
-    'WandBLogger',
-    'RemoteUploaderDownloader',
-    'TensorboardLogger',
-    'CometMLLogger',
-    'ConsoleLogger',
+    'Logger', 'LoggerDestination', 'FileLogger', 'InMemoryLogger', 'ProgressBarLogger', 'WandBLogger',
+    'RemoteUploaderDownloader', 'TensorboardLogger', 'CometMLLogger', 'ConsoleLogger', 'MLFlowLogger'
```
Comment (mvpatel2000) [inline composer/loggers/__init__.py:30]:
```suggestion
    'Logger', 'LoggerDestination', 'FileLogger', 'InMemoryLogger', 'ProgressBarLogger', 'WandBLogger',
    'RemoteUploaderDownloader', 'TensorboardLogger', 'CometMLLogger', 'ConsoleLogger', 'MLFlowLogger',
```
nit: can we re-add trailing comma

File: composer/loggers/logger.py [hunk 2]
```diff
@@ -64,6 +64,7 @@ def log_metrics(self, metrics: Dict[str, float], step: Optional[int] = None) ->
         if step is None:
             step = self._state.timestamp.batch.value
         for destination in self.destinations:
+
             destination.log_metrics(metrics, step)
```
Comment (mvpatel2000) [inline composer/loggers/logger.py:68]:
```suggestion
        for destination in self.destinations:
            destination.log_metrics(metrics, step)
```

Comment (mvpatel2000) [reply composer/loggers/logger.py:68]: was this intentional change?

File: setup.py [hunk 3]
```diff
@@ -194,6 +194,8 @@ def package_files(prefix: str, directory: str, extension: str):
     'onnxruntime>=1.12.1,<2',
 ]
 
+extra_deps['mlflow'] = ['mlflow>=2.0.1']
```
Comment (mvpatel2000) [inline setup.py:197]:
```suggestion
extra_deps['mlflow'] = ['mlflow>=2.0.1,<3.0']
```

**Review by mvpatel2000 — APPROVED** [review]
Nice! LGTM. Would wait to also get review from @dakinggg though -- 2 reviews for this before we ship to customers would be ideal. 

I assume you've tested end-to-end?

File: composer/loggers/mlflow_logger.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+class MLFlowLogger(LoggerDestination):
+    """Log to `MLFlow <https://www.mlflow.org/docs/latest/index.html>`_.
+
+    Args:
+        experiment_name: (str, optional): MLFLow experiment name,
+        run_name: (str, optional): MLFlow run name.
+        tracking_uri (str | pathlib.Path, optional): MLFlow tracking uri, the URI to the
+            remote or local endpoint where logs are stored (If none it is set to `./mlruns`)
+        rank_zero_only (bool, optional): Whether to log only on the rank-zero process
+            (default: ``False``).
```
Comment (dakinggg) [inline composer/loggers/mlflow_logger.py:28]: This is actually set to default to True (and actually the comet logger docstring is wrong also).

File: tests/loggers/test_mlflow_logger.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+    run_cfg = yaml.safe_load(open(str(run_cfg_file_path), 'r'))
+    expected_run_name = mlflow_run_name
+    actual_run_name = run_cfg['run_name']
+    assert actual_run_name == expected_run_name
+
+    test_mlflow_logger.post_close()
+
+    # check run ended
+    assert mlflow.active_run() is None
+    del trainer
```
Comment (dakinggg) [inline tests/loggers/test_mlflow_logger.py:62]: was this necessary for some reason? I don't think it should be, and would like to know if it is.

File: tests/loggers/test_mlflow_logger.py [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+    num_batches = 4
+    eval_interval = '1ba'
+
+    trainer = Trainer(model=SimpleConvModel(),
+                      loggers=test_mlflow_logger,
+                      train_dataloader=DataLoader(RandomImageDataset(size=dataset_size), batch_size),
+                      eval_dataloader=DataLoader(RandomImageDataset(size=dataset_size), batch_size),
+                      max_duration=f'{num_batches}ba',
+                      eval_interval=eval_interval,
+                      algorithms=[SqueezeExcite()])
```
Comment (dakinggg) [inline tests/loggers/test_mlflow_logger.py:81]: Any particular reason to have an algorithm on this test?

File: tests/loggers/test_mlflow_logger.py [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+
+    mlflow_uri = tmp_path / Path('my-test-mlflow-uri')
+    mlflow_exp_name = 'my-test-mlflow-exp'
+    mlflow_run_name = 'my-test-mlflow-run'
+
+    test_mlflow_logger = MLFlowLogger(experiment_name=mlflow_exp_name,
+                                      run_name=mlflow_run_name,
+                                      tracking_uri=mlflow_uri)
+
+    dataset_size = 16
```
Comment (dakinggg) [inline tests/loggers/test_mlflow_logger.py:29]: It doesn't look like this test actually needs a trainer? you can just call init on the logger. the other test tests its integration with the trainer and that metrics get logged.

**Review by dakinggg — APPROVED** [review]
Basically LGTM, left a few comments

File: composer/loggers/mlflow_logger.py [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+
+        if self.experiment_name is None:
+            self.experiment_name = 'my-mlflow-experiment'
+
+        if self._enabled:
+            if self.tracking_uri is not None:
+                mlflow.set_tracking_uri(self.tracking_uri)
+            mlflow.set_experiment(self.experiment_name)
+            mlflow.start_run()
+            mlflow.set_tag('mlflow.runName', self.run_name)
```
Comment (dakinggg) [inline composer/loggers/mlflow_logger.py:72]: It looks like run name is supposed to be passed in as an argument to `start_run`? https://www.mlflow.org/docs/latest/python_api/mlflow.html#mlflow.start_run

File: composer/loggers/mlflow_logger.py (hunk 4 above)
Comment (eracah) [reply composer/loggers/mlflow_logger.py]: fixed

File: tests/loggers/test_mlflow_logger.py (hunk 5 above)
Comment (eracah) [reply tests/loggers/test_mlflow_logger.py]: you mean the del trainer? removed.

File: tests/loggers/test_mlflow_logger.py (hunk 6 above)
Comment (eracah) [reply tests/loggers/test_mlflow_logger.py]: no not really just wanted some more hparam examples. removed.

File: tests/loggers/test_mlflow_logger.py (hunk 7 above)
Comment (eracah) [reply tests/loggers/test_mlflow_logger.py]: good call. I was just too lazy to make a mock logger and state. fixed

File: composer/loggers/mlflow_logger.py (hunk 8 above)
Comment (eracah) [reply composer/loggers/mlflow_logger.py]: they are both valid ways, but you're right this one is simpler. fixed.

**Comment by eracah** [comment]
> Nice! LGTM. Would wait to also get review from @dakinggg though -- 2 reviews for this before we ship to customers would be ideal. Edit: Looks like he's codeowner anyways so its forced
> 
> I assume you've tested end-to-end?

yup. they don't have a native dashboard. it's similar to tensorboard.

**Pull request merged**